rdrf/rdrf/management/commands/migrate_gridfs_django.py

```diff
@@ -0,0 +1,183 @@
+
+    for form_index, form in enumerate(doc.get("forms") or []):
+        for sec_index, section in enumerate(form.get("sections") or []):
+            for cde_index, cde in enumerate(section.get("cdes") or []):
+                gridfs_file_id, filename = get_gridfs_file_id(cde)
+                if gridfs_file_id:
+                    idx = (prefix, form_index, sec_index, cde_index)
+                    path = "%sforms.%d.sections.%d.cdes.%d.value" % idx
+                    context = {
+                        "form_name": form.get("name"),
+                        "section_code": section.get("code"),
+                        "cde_code": cde.get("code"),
+                    }
+                    yield DocUpdate(collection, path, gridfs_file_id, filename, context)
+
+def convert_file(fs, registry, context, gridfs_file_id, filename, dry_run=False):
+    try:
+        logger.info("Finding file %s %s" % (str(gridfs_file_id), filename))
+        data = fs.get(gridfs_file_id)
+        logger.info("Copying %d bytes" % data.length)
+        temp = TemporaryFile(prefix=filename)
+        copyfileobj(data, temp)
+        temp.seek(0)
+        md5 = calc_md5(temp)
+
+        if md5 == data.md5:
+            logger.info("Loading into Django storage md5=%s" % md5)
+            temp.seek(0)
+            cde_file = make_cde_file(temp, filename, registry, **context)
+            if dry_run:
+                return 0
+            if cde_file:
+                cde_file.save()
+                return cde_file.id
+            return None
+        else:
+            logger.warning("Can't convert file: MD5 mismatch")
+            return None
+    except gridfs.NoFile:
+        logger.warning("GridFS file %s does not exist" % str(gridfs_file_id))
+        return None
+    except Exception as e:
+        logger.exception("Problem converting file")
+        return None
+
+def make_cde_file(file_obj, filename, registry, cde_code=None,
+                  form_name=None, section_code=None):
+    if form_name:
+        form = RegistryForm.objects.filter(name=form_name).first()
+        if not form:
+            logger.warning("RegistryForm with name \"%s\" doesn't exist" % form_name)
+    else:
+        form = None
+
+    if section_code:
+        section = Section.objects.filter(code=section_code).first()
+        if not section:
+            logger.warning("Section with code \"%s\" doesn't exist" % section_code)
+
+    cde = CommonDataElement.objects.filter(code=cde_code).first()
+    if not cde:
+        logger.warning("CDE with code \"%s\" doesn't exist" % cde_code)
+        return None
+
+    return CDEFile(registry=registry,
+                   form=form,
+                   section=section,
+                   cde=cde,
+                   item=File(file_obj),
+                   filename=filename or "")
+
+
+def calc_md5(file_obj):
+    h = hashlib.new("md5")
+    while True:
+        buf = file_obj.read(16*1024)
+        if not buf:
+            break
```

Comment:
Second version of iter(), the one with sentinel values, reads better IMHO.
No while True, no break on end value.